input/script_utils.go

```diff
@@ -249,6 +249,109 @@ func SenderHTLCScript(senderHtlcKey, receiverHtlcKey,
+
+	builder := txscript.NewScriptBuilder()
+
+	// The opening operations are used to determine if this is the receiver
+	// of the HTLC attempting to sweep all the funds due to a contract
+	// breach. In this case, they'll place the revocation key at the top of
+	// the stack.
+	builder.AddOp(txscript.OP_DUP)
+	builder.AddOp(txscript.OP_HASH160)
+	builder.AddData(btcutil.Hash160(revocationKey.SerializeCompressed()))
+	builder.AddOp(txscript.OP_EQUAL)
+
+	// If the hash matches, then this is the revocation clause. The output
+	// can be spent if the check sig operation passes.
+	builder.AddOp(txscript.OP_IF)
+	builder.AddOp(txscript.OP_CHECKSIG)
+
+	// Otherwise, this may either be the receiver of the HTLC claiming with
+	// the pre-image, or the sender of the HTLC sweeping the output after
+	// it has timed out.
+	builder.AddOp(txscript.OP_ELSE)
+
+	// We'll do a bit of set up by pushing the receiver's key on the top of
+	// the stack. This will be needed later if we decide that this is the
+	// sender activating the time out clause with the HTLC timeout
+	// transaction.
+	builder.AddData(receiverHtlcKey.SerializeCompressed())
+
+	// Atm, the top item of the stack is the receiverKey's so we use a swap
+	// to expose what is either the payment pre-image or a signature.
+	builder.AddOp(txscript.OP_SWAP)
+
+	// With the top item swapped, check if it's 32 bytes. If so, then this
+	// *may* be the payment pre-image.
+	builder.AddOp(txscript.OP_SIZE)
+	builder.AddInt64(32)
+	builder.AddOp(txscript.OP_EQUAL)
+
+	// If it isn't then this might be the sender of the HTLC activating the
+	// time out clause.
+	builder.AddOp(txscript.OP_NOTIF)
+
+	// We'll drop the OP_IF return value off the top of the stack so we can
+	// reconstruct the multi-sig script used as an off-chain covenant. If
+	// two valid signatures are provided, ten then output will be deemed as
+	// spendable.
+	builder.AddOp(txscript.OP_DROP)
+	builder.AddOp(txscript.OP_2)
+	builder.AddOp(txscript.OP_SWAP)
+	builder.AddData(senderHtlcKey.SerializeCompressed())
+	builder.AddOp(txscript.OP_2)
+	builder.AddOp(txscript.OP_CHECKMULTISIG)
+
+	// Otherwise, then the only other case is that this is the receiver of
+	// the HTLC sweeping it on-chain with the payment pre-image.
+	builder.AddOp(txscript.OP_ELSE)
+
+	// Hash the top item of the stack and compare it with the hash160 of
+	// the payment hash, which is already the sha256 of the payment
+	// pre-image. By using this little trick we're able save space on-chain
+	// as the witness includes a 20-byte hash rather than a 32-byte hash.
+	builder.AddOp(txscript.OP_HASH160)
+	builder.AddData(Ripemd160H(paymentHash))
+	builder.AddOp(txscript.OP_EQUALVERIFY)
+
+	// This checks the receiver's signature so that a third party with
+	// knowledge of the payment preimage still cannot steal the output.
+	builder.AddOp(txscript.OP_CHECKSIG)
+
+	// Close out the OP_IF statement above.
+	builder.AddOp(txscript.OP_ENDIF)
+
+	// 1 block CSV delay.
```

Comment:
can this be simplified by having a subroutine with a boolean arg as to whether or not to add these two ops? same for the receiver script.

Reply:
Yeah as mentioned in the comment above this can be done, but didn't do it to keep script independent . But it's a lot of duplication so maybe better to just do it.